source/common/stats/stats_impl.h

```diff
@@ -473,5 +477,114 @@ class IsolatedStoreImpl : public Store {
   IsolatedStatsCache<Histogram, HistogramImpl> histograms_;
 };
 
+/**
+ * Structure to hold statistical data of histogram.
+ */
+struct HistogramStatistics {
+public:
+  HistogramStatistics(const std::string name, histogram_t* histogram_ptr) : name_(name) {
+    double arr[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
+    double out[9];
+    double* arrptr = arr;
+    double* outptr = out;
+    hist_approx_quantile(histogram_ptr, arrptr, 9, outptr);
+    p0_ = out[0];
+    p25_ = out[1];
+    p50_ = out[2];
+    p75_ = out[3];
+    p90_ = out[4];
+    p95_ = out[5];
+    p99_ = out[6];
+    p999_ = out[7];
+    p100_ = out[8];
+  }
+
+  std::string name() const { return name_; }
+
+  std::string summary() const {
+    return fmt::format("P0: {} , P25: {}, P50: {}, P75: {}, P90: {}, P95: {}, P99: {}, P100: {}",
+                       p0_, p25_, p50_, p75_, p90_, p95_, p99_, p999_, p100_);
+  }
+
+private:
+  const std::string name_;
+  double p0_;
+  double p25_;
+  double p50_;
+  double p75_;
+  double p90_;
+  double p95_;
+  double p99_;
+  double p999_;
+  double p100_;
+};
+
+/**
+ * BaseSink implementation for all Sinks. This mainly  .
+ */
+class BaseSink : public Sink {
+public:
+  void onHistogramComplete(const Histogram& hist, uint64_t value) override {
+    const auto histogram_value = tls_->getTyped<ThreadLocalHistograms>().histograms_.emplace(
+        std::make_pair(hist.name(), hist_alloc()));
+    histogram_t* histogram_ptr = histogram_value.first->second;
+    hist_insert(histogram_ptr, value, 1);
+  }
+
+  std::list<HistogramStatistics> flushHistograms() override {
+    // TODO: Merge with optimized locking
+    std::lock_guard<std::mutex> guard(lock_);
```

Comment:
need to lock the structures as well. for the time being just added this very basic `gaurd` which is not at all sufficient.

Reply:
Per my other comment about whether everyone gets this, especially without optimized RCU type locking, I wonder if we should not enable this for now for statsd? Perhaps that sink should not derive from BaseSink? Or BaseSink should have a boolean as to whether local histograms are kept? Along these lines, should we add an explicit PrometheusSink so we can clearly enable this? Or some other config option?

Reply:
I think having a boolean makes sense at BaseSink. Will make that change and we can disable if needed for statsd Sinks.